File: 01_Acey_Ducey/ruby/aceyducey.rb
########################################################
#
# Acey Ducey
#
# From: BASIC Computer Games (1978)
#       Edited by David Ahl
#
# "This is a simulation of the Acey Ducey card game.
#  In the game, the dealer (the computer) deals two
#  cards face up.  You have an option to bet or not to
#  bet depending on whether or not you feel the next
#  card dealt will have a value between the first two.
#
# "Your initial money is set to $100. The game keeps
#  going on until you lose all your money or interrupt
#  the program.
#
# "The original BASIC program author was Bill Palmby
#  of Prairie View, Illinois."
#
# Ruby port by Christopher Oezbek, 2021
#
# This uses the following techniques:
#
#  - The programm largely consists of a GAME LOOP,
#    which is used to represent one round.
#  - The Kernel function rand(Range) is used to get an
#    Integer in the (inclusive) Range of 2 to 14.
#  - To ensure the user enters a proper Integer
#    via the console, an inline 'rescue' statement is
#    used.
#  - To capture the long text in the introduction, a
#    squiggly HEREDOC string declaration <<~ is used.
#
#
########################################################

puts <<~INSTRUCTIONS
           🂡  ACEY DUCEY CARD GAME 🂱
   CREATIVE COMPUTING - MORRISTOWN, NEW JERSEY

  ACEY-DUCEY IS PLAYED IN THE FOLLOWING MANNER
  THE DEALER (COMPUTER) DEALS TWO CARDS FACE UP
  YOU HAVE AN OPTION TO BET OR NOT BET DEPENDING
  ON WHETHER OR NOT YOU FEEL THE CARD WILL HAVE
  A VALUE BETWEEN THE FIRST TWO.
  IF YOU DO NOT WANT TO BET IN A ROUND, ENTER 0
INSTRUCTIONS

# The player starts with 100$
stake = 100

while true # Game loop
  puts
  puts "YOU NOW HAVE #{stake} DOLLARS."
  puts
  puts "HERE ARE YOUR NEXT TWO CARDS:"

  # Randomly draw two cards and make sure the first card is lower in value than the second
  # Using array destructuring, this sorted array can be assigned to `first_card` and `second_card`
  first_card, second_card = (2...14).to_a.shuffle.pop(2).sort

  # Helper method to convert a numeric card into a String for printing
  def card_name(card)
    case card
    when 11
      "JACK"
    when 12
      "QUEEN"
    when 13
      "KING"
    when 14
      "ACE"
    else
      card
    end
  end

  puts card_name(first_card)
  puts card_name(second_card)
  puts
  puts

  # Loop until the user enters something sensible
  while true
    puts "WHAT IS YOUR BET? ENTER 0 IF YOU DON'T WANT TO BET (CTRL+C TO EXIT)"
    your_bet = Integer(gets.chomp) rescue nil

    if your_bet == nil || your_bet < 0
      puts "PLEASE ENTER A POSITIVE NUMBER"
      puts
      next
    end

    if your_bet > stake
      puts "SORRY, MY FRIEND, BUT YOU BET TOO MUCH."
      puts "YOU HAVE ONLY #{stake} DOLLARS TO BET."
      puts
      next
    end

    break
  end

  if your_bet == 0
    puts "CHICKEN!!"
    next
  end

  puts "THANK YOU! YOUR BET IS #{your_bet} DOLLARS."
  puts ""
  puts "THE THIRD CARD IS:"
  third_card = rand(2..14)
  puts card_name(third_card)
  puts

  if first_card <= third_card && third_card <= second_card
    puts "YOU WIN!!!"
    stake += your_bet
  else
    puts "SORRY, YOU LOSE"
    stake -= your_bet
  end

  if stake == 0
    puts
    puts "SORRY, FRIEND, BUT YOU BLEW YOUR WAD."
    puts

    puts "TRY AGAIN? (YES OR NO)"
    if gets.chomp.downcase.start_with? 'y'
      # Reset cash to 100
      stake = 100
    else
      puts "O.K., HOPE YOU HAD FUN!"
      exit
    end
  end
end

File: 01_Acey_Ducey/lua/acey_ducey.lua
print [[
            ACEY DUCEY CARD GAME
 CREATIVE COMPUTING  MORRISTOWN, NEW JERSY



ACEY-DUCEY IS PLAYED IN THE FOLLOWING MANNER
THE DEALER (COMPUTER) DEALS TWO CARDS FACE UP
YOU HAVE AN OPTION TO BET OR NOT BET DEPENDING
ON WHETHER OF NOT YOU FEEL THE CARD WILL HAVE
A VALUE BETWEEN THE FIRST TWO.
IF YOU DO NOT WANT TO BET INPUT A 0]]

local starting_cash = 100

local card_names = {[11] = "JACK", [12] = "QUEEN", [13] = "KING", [14] = "ACE"}
for i = 2, 10 do
  card_names[i] = string.format(" %d", i)
end

function play_round(player_cash, skip_total)
  assert(player_cash > 0, "can't play with nothing to bet")

  if not skip_total then
    print(string.format("YOU NOW HAVE %u DOLLARS", player_cash))
  end

  print "\nHERE ARE YOUR NEXT TWO CARDS"

  local first_card, second_card;

  repeat
    first_card = math.random(2, 14)
    second_card = math.random(2, 14)
  until first_card < second_card

  print(card_names[first_card])
  print(card_names[second_card])
  print("")

  local bet = get_bet(player_cash)

  if bet == 0 then
    print "CHICKEN!!"
    return play_round(player_cash, "skip total")
  end

  local third_card = math.random(2, 14)
  print(card_names[third_card])

  if first_card < third_card and third_card < second_card then
    print "YOU WIN!!!"
    return play_round(player_cash + bet)
  end

  print "SORRY, YOU LOSE"
  if bet < player_cash then
    return play_round(player_cash - bet)
  end

  print "SORRY, FRIEND BUT YOU BLEW YOUR WAD"
  io.write "TRY AGAIN (YES OR NO)? "
  local keep_playing = io.read("l")
  if keep_playing:upper():match("%f[%a]YES%f[^%a]") then
    return play_round(starting_cash)
  end

  print "OK HOPE YOU HAD FUN"
end

function get_bet(player_cash)
  assert(player_cash > 0, "can't play with nothing to bet")

  io.write("WHAT IS YOUR BET? ")
  local input = io.read("l")
  if not input then
    print "GOODBYE"
    os.exit(0)
  end

  local digits = input:match("(%d+)")
  if not digits then
    print "I DON'T UNDERSTAND THAT NUMBER"
    return get_bet(player_cash)
  end

  local bet = tonumber(digits)
  assert(bet, "pattern matched something that doesn't convert to number")

  if bet > player_cash then
    print "SORRY, MY FRIEND BUT YOU BET TOO MUCH"
    print(string.format("YOU HAVE ONLY %u DOLLARS TO BET", player_cash))
    return get_bet(player_cash)
  end

  assert(0 <= bet and bet <= player_cash, "invalid bet")
  return bet
end

play_round(starting_cash)
